src/checks/table_data_completeness.py

```diff
@@ -1,27 +1,31 @@
-import shapefile
+from geopandas import GeoDataFrame
+
+from src.utils import CheckReturnList
 
 EMPTY_VALUES = (None, "null", "")
 
 
-def table_data_completeness(shapefile_path: str) -> float:
-    """Check completeness of a shapefile's table data.
+def main(iso3: str, gdfs: list[GeoDataFrame]) -> CheckReturnList:
+    """Check completeness of an admin boundary's table data.
 
-    Iterates through a shapefile's records and checks for values that are None,
-    null or empty strings.
+    Iterates through a file's records and checks for values that are None,
+    null or empty strings. Uses vectorized pandas for speed.
 
     Args:
-        shapefile_path - a string representing the shapefile to analyse.
+        iso3: ISO3 code of the current location being checked.
+        gdfs: List of GeoDataFrames, with the item at index 0 corresponding to admin
+        level 0, index 1 to admin level 1, etc.
 
     Returns:
-         float between 0 and 1. 1 indicates complete table data (i.e. no
-         missing data). A value of 0 indicates all table data is missing.
+        List of check rows to be outputed as a CSV.
     """
-    complete_records, missing_records = 0, 0
-    sf = shapefile.Reader(shapefile_path)
-    for record in sf.iterRecords():
-        for field in record:
-            if field in EMPTY_VALUES:
-                missing_records += 1
-            else:
-                complete_records += 1
-    return complete_records / (complete_records + missing_records)
+    check_results = []
+    for admin_level, gdf in enumerate(gdfs):
+        row = {
+            "iso3": iso3,
+            "level": admin_level,
+            "total_number_of_records": gdf.size,
+            "number_of_missing_records": gdf.isin(EMPTY_VALUES).sum().sum(),
```

Comment:
Nice suggestion with `.astype(int)`. Added. 

With that missing value, I don't know what values we should expect to denote missing. Do you? Your suggestions there are a good improvement, but do we need to handle anything more exotic, e.g. -9999 as denoting missing values?

Reply:
I don't think so, from what I saw in the document there are other checks that cover that such as feature 2.2.6. The kinds of errors I'd expect are things like data originating from a Shapefile with Arabic characters being saved to LATIN1 instead of UTF-8 and having names that are all `??????`.

Reply:
I see there are a lot of matches for empty strings comparing the CSV output with and without that part. What I don't know is whether there are also empty cells with `" "` or `"  "`, etc. I might add something like a `gdf.str.strip() == ""` check. (can't remember the exact pandas syntax that would do that)

Reply:
I couldn't get this down to anything neater than an apply, which is a little clunky, but works

Reply:
Here's a version that's a bit cleaner: 
```py
"number_of_missing_records": (gdf.isna().stack() | gdf.stack().str.isspace()).sum(),
```